challenger.go

```diff
@@ -50,13 +69,129 @@ func NewLndChallenger(cfg *authConfig, genInvoiceReq InvoiceRequestGenerator) (
 	return &LndChallenger{
 		client:        client,
 		genInvoiceReq: genInvoiceReq,
+		invoiceStates: make(map[lntypes.Hash]lnrpc.Invoice_InvoiceState),
+		quit:          make(chan struct{}),
 	}, nil
 }
 
+// Start starts the challenger's main work which is to keep track of all
+// invoices and their states. For that the backing lnd node is queried for all
+// invoices on startup and the a subscription to all subsequent invoice updates
+// is created.
+func (l *LndChallenger) Start() error {
+	// These are the default values for the subscription. In case there are
+	// no invoices yet, this will instruct lnd to just send us all updates.
+	// If there are existing invoices, these indices will be updated to
+	// reflect the latest known invoices.
+	addIndex := uint64(0)
+	settleIndex := uint64(0)
+
+	// Get a list of all existing invoices on startup and add them to our
+	// cache. We need to keep track of all invoices, even quite old ones to
+	// make sure tokens are valid. But to save space we only keep track of
+	// an invoice's state.
+	invoiceResp, err := l.client.ListInvoices(
+		context.Background(), &lnrpc.ListInvoiceRequest{
+			NumMaxInvoices: math.MaxUint64,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	// Advance our indices to the latest known one so we'll only receive
+	// updates for new invoices and/or newly settled invoices.
+	l.invoicesMtx.Lock()
+	for _, invoice := range invoiceResp.Invoices {
```

Comment:
I think it should be safe to ignore known _expired_ invoices here. If we want to cut things down even further, we can use the current LSAT price as well (ignore invoices that don't match value-wise).